src/WebJobs.Script/Scale/HostPerformanceManager.cs

```diff
@@ -27,20 +39,99 @@ public HostPerformanceManager(IEnvironment environment, IOptions<HostHealthMonit
 
             _environment = environment;
             _healthMonitorOptions = healthMonitorOptions;
+            _logger = logger;
+            _serviceProvider = serviceProvider;
+            _processMonitor = new ProcessMonitor(Process.GetCurrentProcess().Id, environment);
+
+            _processMonitor.Start();
         }
 
-        public virtual bool IsUnderHighLoad(Collection<string> exceededCounters = null, ILogger logger = null)
+        /// <summary>
+        /// Check sandbox enforced performance counters
+        /// </summary>
+        public virtual bool PerformanceCountersExceeded(Collection<string> exceededCounters = null, ILogger logger = null)
         {
             var counters = GetPerformanceCounters(logger);
             if (counters != null)
             {
-                return IsUnderHighLoad(counters, exceededCounters, _healthMonitorOptions.Value.CounterThreshold);
+                return PerformanceCounterThresholdsExceeded(counters, exceededCounters, _healthMonitorOptions.Value.CounterThreshold);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Check both sandbox enforced performance counters as well as process level thresholds
+        /// like CPU, memory, etc.
+        /// </summary>
+        public virtual async Task<bool> IsUnderHighLoad(Collection<string> exceededCounters = null, ILogger logger = null)
+        {
+            return PerformanceCountersExceeded(exceededCounters, logger) ||
+                await ProcessThresholdsExceeded(exceededCounters, logger);
+        }
+
+        private async Task<bool> ProcessThresholdsExceeded(Collection<string> exceededCounters = null, ILogger logger = null)
+        {
+            var hostProcessStats = _processMonitor.GetStats();
+            if (hostProcessStats.CpuLoadHistory.Any())
+            {
+                string formattedLoadHistory = string.Join(",", hostProcessStats.CpuLoadHistory);
+                logger?.LogDebug($"[HostMonitor] Host process stats: EffectiveCores={_environment.GetEffectiveCoresCount()}, CpuLoadHistory=({formattedLoadHistory}), AvgLoad: {Math.Round(hostProcessStats.CpuLoadHistory.Average())}, MaxLoad: {Math.Round(hostProcessStats.CpuLoadHistory.Max())}");
+            }
+
+            double workersCpuAverage = 0;
```

Comment:
If the average for hostProcess is already above the threshold, you probably want to skip calculating the workers cpu averages and return the true

Reply:
I could do that, but the calculations aren't expensive, and doing them emits useful heartbeat status info that'll be good to have in the logs.